### PR TITLE
[Web] Add List View to Unified Resources

### PR DESCRIPTION
**Pull request by rudream**
## Purpose

Adds a list view option to the Unified Resources page as an alternative to the card view.

Also changes the hover state for the card view to use a shadow.

`e` counterpart: https://github.com/gravitational/teleport.e/pull/2619

[Figma design](https://www.figma.com/file/JKd0Bgs30AnNfNCLUuzzc2/Unified-Resources?type=design&node-id=2059-46350&mode=design&t=KjCofsBgVwC0TvXF-0)

## Demo

image

## TODO:

- [x] Add view mode preference to user preferences.
- [x] Add skeleton loader. https://github.com/gravitational/teleport/pull/34381
- [ ] Force card view when viewport is narrow.

changelog: Add list view to resources page

### CODE REVIEW

**Comment by gzdunek** [comment]
Something is causing the the horizontal scrollbar blinking (Firefox):

https://github.com/gravitational/teleport/assets/20583051/858974f4-ac05-4ab8-9015-20fec0e28b68

**Comment by zmb3** [comment]
What do you think about putting a tooltip over the labels icon that says "Show Labels" (similar to how we show the "Click to Copy" tooltip on the copy button)?

**Comment by rudream** [comment]
> Something is causing the the horizontal scrollbar blinking (Firefox)

@gzdunek Good catch, looks like this was being caused because the `box-shadow` on hover was increasing the width by a few pixels. I added a workaround using a pseudo-element to prevent this.

File: web/packages/shared/components/ToolTip/HoverTooltip.tsx [hunk 1]
```diff
@@ -0,0 +1,81 @@
+/**
```
Comment (avatus) [inline web/packages/shared/components/ToolTip/HoverTooltip.tsx:1]: thanks for moving this here

**Comment by avatus** [comment]
This looks/works great. My only concern is leaving out the user preferences part because until that goes in, a user who would want this view would have to click the list view button every time they load the page (could lead to more frustration than not having it at all).

File: web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.tsx [hunk 2]
```diff
@@ -250,17 +223,17 @@ export function ResourceCard({
...
                   </Text>
                 </SingleLineBox>
               )}
-              {description.secondary && (
-                <SingleLineBox ml={2} title={description.secondary}>
+              {addr && (
+                <SingleLineBox ml={2} title={addr}>
                   <Text typography="body2" color="text.muted">
-                    {description.secondary}
+                    {addr}
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.tsx:236]: In a list view we show the description, but here we don't. Is this intentional? 
An idea that comes to my mind is to show one or another (`description || addr`, but I don't know which one is actually more important). I'd confirm it with design.

File: web/packages/shared/components/UnifiedResources/shared/utils.ts [hunk 3]
```diff
@@ -14,133 +14,116 @@
...
-  SecondaryIcon: typeof Icon;
-  ActionButton: React.ReactElement;
-}
+  UnifiedResourceKube,
+  UnifiedResourceUserGroup,
+  SharedUnifiedResource,
+} from '../types';
 
-export function makeUnifiedResourceCardNode(
+export function makeUnifiedResourceItemNode(
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/shared/utils.ts:41]: These functions no longer have to be exported.

File: web/packages/shared/components/UnifiedResources/shared/utils.ts [hunk 4]
```diff
@@ -14,133 +14,116 @@
...
   UnifiedResourceUi,
-  UnifiedResourceDatabase,
+  UnifiedResourceNode,
   UnifiedResourceApp,
-  UnifiedResourceUserGroup,
+  UnifiedResourceDatabase,
   UnifiedResourceDesktop,
-} from './types';
-
-export interface UnifiedResourceCard {
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/shared/utils.ts:42]: Actually, it made sense to me to keep the the type here, collocated with functions that used it. https://www.totaltypescript.com/where-to-put-your-types-in-application-code 

I'd treat `types.ts` as the last resort option, for situations where type is imported by multiple files or importing a type would cause an import loop.
In our case, the type is used only in this file.

File: web/packages/shared/components/UnifiedResources/shared/utils.ts
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/shared/utils.ts:1]: `utils.ts` name is not very good at describing what the file contains (and it encourages to put unrelated stuff to that file).
Maybe we could call it `viewItemsFactory.ts`? I'd then rename `UnifiedResourceItem` type to `UnifiedResourceViewItem` to indicate that this an internal type used to generate views.

File: web/packages/shared/components/UnifiedResources/shared/index.ts
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/shared/index.ts:1]: TBH I'd remove this file. 

> Index.js is not about having to type out slightly shorter file names.
> 
> It is about encapsulation.
> 
> The idea is that your directories should have files and classes or services or whatever you want it have but everything this is public should be exported or exposed in the index.js
> 
> The idea is to never require a file from inside another directory directly as whatever it exposes should be that "package" 's internals and are private.
> 
> Of course this is not enforced in JavaScript, but it is important for writing clean, well organised code.

https://www.reddit.com/r/javascript/comments/7rv221/comment/dszx99t/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Since we are in a private directory, there is no need to re-export its stuff to the parent, which is still private. This file mainly adds "noise".

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+          <Box
+            css={`
+              align-self: start;
+            `}
+          >
+            {hovered && <CopyButton name={name} />}
+          </Box>
+        </Flex>
+
+        {/* type */}
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:176]: When the viewport is narrow, the resource type looks like this:
image

I see `ellipsis` on the container so I guess it was supposed to behave differently.

**Review by gzdunek — COMMENTED** [review]
Second pass

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+      } else {
+        setIsDescOverflowed(false);
+      }
+    });
+
+    observer.observe(innerContainer.current);
+    return () => {
+      observer.disconnect();
+    };
+  });
```
[the remark this answers is not in the document]
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:83]: I really like this idea, cleaned up a lot of the hard-to-read logic in the component. Thanks!

File: web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.tsx (hunk 2 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.tsx:236]: This is intentional and I discussed it with design, the card layout doesn't allow for adding a description. From a UX perspective I'm not sure it's a good idea to have some cards show a description and others show an address in the same place as it could be confusing, I'll ask to see what Kenny thinks.

File: web/packages/shared/components/UnifiedResources/shared/utils.ts (hunk 3 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/shared/utils.ts:41]: We still use them in `ResourceCard.story.tsx` and `ResourceListItem.story.tsx` in order to convert our `Node`, `App`, etc. fixtures to Unified Resource items.

File: web/packages/shared/components/UnifiedResources/shared/utils.ts (hunk 4 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/shared/utils.ts:42]: After my new changes we do actually use this type in another place (in `ResourceViewProps`). Thanks for the suggestion! I'll keep this in mind next time I'm working with `types.ts` files

File: web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.tsx (hunk 2 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/CardsView/ResourceCard.tsx:221]: This is intentional, it's because a description doesn't work well with the card's limited layout and space. I will keep the card layout the way it currently is in `master`.

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 7]
```diff
@@ -0,0 +1,321 @@
...
+            <Text fontSize="14px" fontWeight={300} color="text.muted">
+              {addr}
+            </Text>
+          </HoverTooltip>
+        </Box>
+
+        {/* show labels button */}
+        {labels.length > 0 && (
+          <HoverTooltip
+            tipContent={showLabels ? 'Hide Labels' : 'Show Labels'}
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:176]: `Show labels`, `Hide labels`? This will align with other tooltips (`Sort by`).

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+            css={`
+              grid-area: labels-btn;
+            `}
+          >
+            <ShowLabelsButton
+              size={1}
+              onClick={() => setShowLabels(!showLabels)}
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:183]:
```suggestion
              onClick={() => setShowLabels(prevState => !prevState)}
```

File: web/packages/shared/components/UnifiedResources/UnifiedResources.tsx [hunk 9]
```diff
@@ -291,7 +279,17 @@ export function UnifiedResources(props: UnifiedResourcesProps) {
...
+      ...unifiedResourcePreferences,
+      defaultTab: value,
+    });
+  };
+
+  const selectViewMode = (viewMode: UnifiedViewModePreference) => {
+    updateUnifiedResourcesPreferences({
+      ...unifiedResourcePreferences,
+      viewMode,
+    });
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:292]: This is great 👏 
However, now we have some inconsistency: `viewMode` state is kept outside, but information if pinned tab is active is kept in params and sent externally via `updateUnifiedResourcesPreferences`.
It seems to me that it would be good to have a single way of updating both properties. I made an attempt to refactor this, you can check out the draft https://github.com/gravitational/teleport/pull/34394.
It's not perfect, but we should refactor `useUrlFiltering` first.
I also improved naming and fixed imports (`shared` and `teleterm` shouldn't import from `teleport`).

File: web/packages/shared/components/UnifiedResources/types.ts [hunk 10]
```diff
@@ -100,3 +104,84 @@ export type UnifiedResourcesQueryParams = {
...
+  pinned: boolean;
+  pinningSupport: PinningSupport;
+};
+
+// Props that are needed for the Card view.
+// The reason we need this separately defined is because unlike with the list view, what we display in the
+// description sections of a card varies based on the type of its resource. For example, for applications,
+// instead of showing the `Application` type under the name like we would for other resources, we show the description.
+type CardViewSpecificProps = {
+  primaryDesc?: string;
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/types.ts:154]: `primaryDescription`? In the `ListViewSpecificProps` type we have `description`.

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+          <Box
+            css={`
+              align-self: start;
+            `}
+          >
+            {hovered && <CopyButton name={name} ml={1} />}
+          </Box>
+        </Flex>
+
+        {/* type */}
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:126]: You know what would be cool? If we could show the full type (without ellipsis) if there is no address.
image
Now we are wasting space a bit :)

**Review by gzdunek — COMMENTED** [review]
We are almost there!

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 12]
```diff
@@ -0,0 +1,321 @@
...
+    <RowContainer
+      onMouseEnter={() => setHovered(true)}
+      onMouseLeave={() => setHovered(false)}
+    >
+      <RowInnerContainer alignItems="start" pinned={pinned} selected={selected}>
+        {/* checkbox */}
+        <HoverTooltip
+          css={`
+            grid-area: checkbox;
+          `}
```
[the remark this answers is not in the document]
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:62]: > You can use `css` only on elements than you can style via `className`

I didn't know this, thanks!

File: web/packages/shared/components/UnifiedResources/UnifiedResources.tsx (hunk 9 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:292]: Agreed, though I think we can do the refactoring after this is merged since this PR is already large enough and this is just refactoring which doesn't affect UX.

Comment (rudream) [reply web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:292]: Agreed, though I think we should do the refactoring after this is merged since this PR is already large enough and this is just refactoring which doesn't affect UX.

File: web/packages/shared/components/UnifiedResources/types.ts (hunk 10 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/types.ts:154]: We do this because `primaryDescription` can be any arbitrary text we decide it to be and is just defining "the main text under the name." Whereas for the list view, `description` represents the actual `description` field in the resource that's returned from the backend.

Comment (rudream) [reply web/packages/shared/components/UnifiedResources/types.ts:154]: I named it this because `primaryDescription` can be any arbitrary text we decide it to be and is essentially just describing the main text under the name in the card. This is essentially the same way we named it before (`description.primary`).

Whereas for the list view, `description` directly represents the actual `description` field in the resource that's returned from the backend, and we don't arbitrarily populate it.

File: web/packages/shared/components/UnifiedResources/UnifiedResources.tsx (hunk 9 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/UnifiedResources.tsx:292]: Agreed, though I think we should do the refactoring after this is merged since this PR is already large enough and this doesn't affect UX.

**Review by avatus — APPROVED** [review]
Looks and works great to me. 👍 , after @gzdunek comments are taken care of

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 13]
```diff
@@ -132,55 +130,57 @@ export function ResourceListItem({
...
+        <HoverTooltip
+          tipContent={addr}
+          showOnlyOnOverflow
           css={`
             grid-area: address;
             overflow: hidden;
             text-overflow: ellipsis;
             white-space: nowrap;
+            // If the labels button isn't showing, let this column take up the extra space.
+            ${!showLabelsButton ? 'grid-column-end: labels-btn;' : ''}
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:165]: Nice 👍

File: web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx [hunk 14]
```diff
@@ -132,55 +130,57 @@ export function ResourceListItem({
             overflow: hidden;
             text-overflow: ellipsis;
             white-space: nowrap;
+            // If there is no address, let this column take up the extra space.
+            ${!addr ? 'grid-column-end: address;' : ''}
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/ListView/ResourceListItem.tsx:134]: This is great, but maybe we could also check for `showLabelsButton`?
And make this wider if this is not showing. The condition would be more complex, perhaps it should be extracted to a function.
image

File: web/packages/shared/components/UnifiedResources/shared/PinButton.tsx [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    <PushPin size="small" />
+  );
+
+  return (
+    // This is wrapped in a Box so that we can disable the transition for showing/hiding the button while not
+    // interfering with the ButtonIcon's transitions.
+    <Box
+      css={`
+        // dont display but keep the layout
+        opacity: ${shouldShowButton ? '1' : '0'};
```
Comment (gzdunek) [inline web/packages/shared/components/UnifiedResources/shared/PinButton.tsx:63]: I think we can go back to 
```ts
visibility: ${shouldShowButton ? 'visible' : 'hidden'};
```
Now when the `PinButton` is hidden, hovering over it changes the cursor to pointer.

**Review by gzdunek — APPROVED** [review]
I left two small comments.

🚢 it!

File: web/packages/shared/components/UnifiedResources/shared/PinButton.tsx (hunk 15 above)
Comment (rudream) [reply web/packages/shared/components/UnifiedResources/shared/PinButton.tsx:63]: I had done this because a `transition` on `visibility` just delays it, which means that when you hovered off of a row the pin button wouldn't disappear instantly. I've found a workaround that works better now
